From patchwork Mon Sep 24 18:05:43 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10612765
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 1EFC6913
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 24 Sep 2018 18:06:03 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 112E52A361
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 24 Sep 2018 18:06:03 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 0594B2A39C; Mon, 24 Sep 2018 18:06:03 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 9D8C12A39D
	for <patchwork-kvm@patchwork.kernel.org>;
 Mon, 24 Sep 2018 18:06:02 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1731606AbeIYAJX (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Mon, 24 Sep 2018 20:09:23 -0400
Received: from mail-ua1-f74.google.com ([209.85.222.74]:34156 "EHLO
        mail-ua1-f74.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1726139AbeIYAJW (ORCPT <rfc822;kvm@vger.kernel.org>);
        Mon, 24 Sep 2018 20:09:22 -0400
Received: by mail-ua1-f74.google.com with SMTP id f12-v6so1352982uan.1
        for <kvm@vger.kernel.org>; Mon, 24 Sep 2018 11:06:00 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:in-reply-to:message-id:mime-version:references:subject:from:to
         :cc;
        bh=JWX8M36riYF1JURcP6Tkf0ew5YFPhWo3CVo61z2c3Ho=;
        b=NMra/E10+cw5x4qIm5jDPPn4neYDdkOwjIjYjr6w8t2DyC7qNfdgCBwnZHRMj4e0uB
         kDuCmSZiYO/wB0BPBgzXZmRBgcZcL3GAeD1zpHBSqsHfl81/wNbiEW04NH1Hx86C5Rb1
         3C2xmtCXf+NgYU0VWgJzJM2XTgtHX3jGztmlFvR2pZJ3T0BxW6mZm149clAJHlzSlbtI
         j/VUgY+xTzCJP7Ql19ooXxNbSp8mVJBO+8SMIoAtcegxTs2vmsndeM/BTbcJzpROo4Sm
         cyA77jkwWghAWqBq9SWuU7x1v87t4BMMn4qwBM1HGymkl7+eUpFVBOkX9m9VUYC/blGn
         GtVw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:in-reply-to:message-id:mime-version
         :references:subject:from:to:cc;
        bh=JWX8M36riYF1JURcP6Tkf0ew5YFPhWo3CVo61z2c3Ho=;
        b=C1oG9dfW+I57sAt7XHlgvn4ZMd52/4XAy2klL1ikJgimHXTiE94RJcmCNGG10JKEAt
         vLgIvXP0F/P97C1B6LYHRQO4YU5CGN8Ue+7nxY37/RGMYwFporf/bKZ7oEgi/kfvfC0j
         klHZ4OSAFwTaKReyTsoLOp+sP9mooYrVPriYsYkeYZsuZcMvzLsHBpIv4aFSr6S7MU+R
         OR+qeDXF3fZS+7q8Pw6CrqRcDNPvfqH/Av8FWyQpKFmt1/K8dyZtWNiXATu/gpFO+FYf
         SVSME9g8ZMRQe52EclGi+vhQLHfalJ7A+2AzpWeHVxnITiV/Luo+Xzz095VML2Yf4IsN
         xFMQ==
X-Gm-Message-State: ABuFfoiXCUhHBRXkJdqKoDnsSOa3ymGHVd2zntMzVEjxyKxVaXMI/g+R
        o555EZBJG3O0O0093O8b3p4715y25gW4/ZIWC/emfslrGUm+rImopDv/4DT53BVUJITdVdPlxAN
        GaJU85tmO6zpNT7BeXW6GDlfHr1U6OT8Hh/KYNvUojMpNQ/N746vzRt7KejJmaT0=
X-Google-Smtp-Source: 
 ACcGV62LlCdB1HSb8yN44ssxZMRL/n7nhzmZBjS6dx48nvVaCp8/wjTCITJbPTSYlzM/7wHfTP0u6OcXaK6lnQ==
X-Received: by 2002:a1f:bd4a:: with SMTP id n71-v6mr13176vkf.0.1537812360211;
 Mon, 24 Sep 2018 11:06:00 -0700 (PDT)
Date: Mon, 24 Sep 2018 11:05:43 -0700
In-Reply-To: <1537807916.8731.10.camel@intel.com>
Message-Id: <20180924180543.227780-1-jmattson@google.com>
Mime-Version: 1.0
References: <1537807916.8731.10.camel@intel.com>
X-Mailer: git-send-email 2.19.0.444.g18242da7ef-goog
Subject: [PATCH v2] KVM: nVMX: Unrestricted guest mode requires EPT
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org,
        Sean Christopherson <sean.j.christopherson@intel.com>
Cc: Andrew Thornton <andrewth@google.com>,
        Peter Shier <pshier@google.com>,
        Jim Mattson <jmattson@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

As specified in Intel's SDM, do not allow the L1 hypervisor to launch
an L2 guest with the VM-execution controls for "unrestricted guest" or
"mode-based execute control for EPT" set and the VM-execution control
for "enable EPT" clear.

Note that the VM-execution control for "mode-based execute control for
EPT" is not yet virtualized by kvm.

Reported-by: Andrew Thornton <andrewth@google.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
Reviewed-by: Sean Christopherson <sean.j.christopherson@intel.com>
Reviewed-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/include/asm/vmx.h |  1 +
 arch/x86/kvm/vmx.c         | 24 ++++++++++++++++++++++++
 2 files changed, 25 insertions(+)

diff --git a/arch/x86/include/asm/vmx.h b/arch/x86/include/asm/vmx.h
index 9527ba5d62da..665632a4b54b 100644
--- a/arch/x86/include/asm/vmx.h
+++ b/arch/x86/include/asm/vmx.h
@@ -78,6 +78,7 @@
 #define SECONDARY_EXEC_RDSEED_EXITING		0x00010000
 #define SECONDARY_EXEC_ENABLE_PML               0x00020000
 #define SECONDARY_EXEC_XSAVES			0x00100000
+#define SECONDARY_EXEC_MODE_BASED_EPT_EXEC	0x00400000
 #define SECONDARY_EXEC_TSC_SCALING              0x02000000
 
 #define PIN_BASED_EXT_INTR_MASK                 0x00000001
diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 06412ba46aa3..b78607dd113c 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -11775,6 +11775,24 @@ static int nested_vmx_check_pml_controls(struct kvm_vcpu *vcpu,
 	return 0;
 }
 
+static int nested_vmx_check_unrestricted_guest_controls(struct kvm_vcpu *vcpu,
+							struct vmcs12 *vmcs12)
+{
+	if (nested_cpu_has2(vmcs12, SECONDARY_EXEC_UNRESTRICTED_GUEST) &&
+	    !nested_cpu_has_ept(vmcs12))
+		return -EINVAL;
+	return 0;
+}
+
+static int nested_vmx_check_mode_based_ept_exec_controls(struct kvm_vcpu *vcpu,
+							 struct vmcs12 *vmcs12)
+{
+	if (nested_cpu_has2(vmcs12, SECONDARY_EXEC_MODE_BASED_EPT_EXEC) &&
+	    !nested_cpu_has_ept(vmcs12))
+		return -EINVAL;
+	return 0;
+}
+
 static int nested_vmx_check_shadow_vmcs_controls(struct kvm_vcpu *vcpu,
 						 struct vmcs12 *vmcs12)
 {
@@ -12397,6 +12415,12 @@ static int check_vmentry_prereqs(struct kvm_vcpu *vcpu, struct vmcs12 *vmcs12)
 	if (nested_vmx_check_pml_controls(vcpu, vmcs12))
 		return VMXERR_ENTRY_INVALID_CONTROL_FIELD;
 
+	if (nested_vmx_check_unrestricted_guest_controls(vcpu, vmcs12))
+		return VMXERR_ENTRY_INVALID_CONTROL_FIELD;
+
+	if (nested_vmx_check_mode_based_ept_exec_controls(vcpu, vmcs12))
+		return VMXERR_ENTRY_INVALID_CONTROL_FIELD;
+
 	if (nested_vmx_check_shadow_vmcs_controls(vcpu, vmcs12))
 		return VMXERR_ENTRY_INVALID_CONTROL_FIELD;
 
